### PR TITLE
samples: mesh: removed bug which is cause of divide by zero exception [1.14 milestone]

### PR DESCRIPTION
**Pull request by vikrant8052**
If Client send transition time values as 0x40 or 0x80 or 0xC0 then
this was  the cause behind the divide by zero exception during calculation
of transition->counter. This PR will solve this bug.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

### CODE REVIEW

**Review by pabigot — COMMENTED** [review]
Would it not make more sense to have `tt_values_calculator` return a value that indicates whether a valid transition was specified (one where `steps_multiplier` is non-zero), and only perform the state adjustments in that case?

It seems wrong to make the multiple call sites responsible for decoding the `tt` field to detect this situation.

**Comment by vikrant8051** [comment]
@jhedberg Please merged it.

**Comment by jhedberg** [comment]
> @jhedberg Please merged it.

I've been waiting for some response from @pabigot if he agrees with your assessment of his concern.

**Comment by pabigot** [comment]
And I was waiting for @jhedberg to express an opinion, since I haven't even tried this example and am not familiar with the protocol.  This was just one of the random PRs I looked at to see how Zephyr does things, and I saw code I thought was difficult to maintain.

No, I don't agree.  `tt_values_calculator` appears to be responsible for decoding the transition time and delay.  Lifting part of that decoding out to every call site does not make sense.  Nor do I believe that a break out of the function to is a heavy burden.

I would prefer something like what's below.  However, I'm not going to test this, so I can't submit a competing proposal.  Nor am I blocking this, because it's not something I care that deeply about.  So it still needs an approving review no matter how it's done.

```
--- a/samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/transition.c
+++ b/samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/transition.c
@@ -84,12 +84,15 @@ static void bound_states_transition_type_reassignment(u8_t type)
        }
 }
 
-static void tt_values_calculator(struct transition *transition)
+static book tt_values_calculator(struct transition *transition)
 {
        u8_t steps_multiplier, resolution;
 
        resolution = (transition->tt >> 6);
        steps_multiplier = (transition->tt & 0x3F);
+       if (steps_multiplier == 0) {
+               return false;
+       }
 
        switch (resolution) {
        case 0: /* 100ms */
@@ -113,6 +116,7 @@ static void tt_values_calculator(struct transition *transition)
        }
 
        ptr_counter = &transition->counter;
+       return true;
 }
 
 void onoff_tt_values(struct generic_onoff_state *state, u8_t tt, u8_t delay)
@@ -122,9 +126,7 @@ void onoff_tt_values(struct generic_onoff_state *state, u8_t tt, u8_t delay)
        state->transition->tt = tt;
        state->transition->delay = delay;
 
-       if ((tt & 0x3F) != 0) {
-               tt_values_calculator(state->transition);
-       } else {
+       if (!tt_values_calculator(state->transition)) {;
                return;
        }
 
```
with the same rework at the other five call sites.

**Comment by jhedberg** [comment]
@pabigot it was perhaps bit of a lazy approach from me since I didn't at that point find the time to take careful look at this, sorry about that. Anyway, what @pabigot is suggesting does sound like cleaner solution to me. @vikrant8051 could you implement that change, please (unless you see some reason why it's absolutely the wrong thing to me - from your comment I got the impression that you don't think either option is "wrong").

**Comment by vikrant8051** [comment]
@jhedberg 
Ok, Please consider now latest amendments.

@pabigot Thanks for your review. If time allow you kindly go through with entire App & help me/us to
improve it further.

**Comment by pabigot** [comment]
On the surface, the new version looks like it went in exactly the opposite direction from what I was suggesting: instead of decoding one value in five places, you now decode two values in five places, and store the values in global variables so they can be fetched from the one place where they're needed.

I don't have any other feedback.

**Comment by jhedberg** [comment]
> @jhedberg Ok, Please consider now latest amendments.

Like @pabigot already mentioned, it doesn't look anything similar to the suggested change.

**Comment by vikrant8051** [comment]
But he is not denied it. Is he ?

**Comment by vikrant8051** [comment]
I wanna just avoid additional function call that is why choose current
approach.

**Comment by jhedberg** [comment]
> I wanna just avoid additional function call that is why choose current
approach.

@vikrant8051 I seriously doubt that there's a real issue with the extra function call. Instead, your latest proposal of avoiding it is just making things worse by creating global variables for state that is only relevant for the current call stack.

If you have any similar issues (global variables for local state) in your app, please fix those in separate PRs (*don't* add more commits to this PR like you've done in the past!)

**Comment by vikrant8051** [comment]
It's done !!

**Review by pabigot — APPROVED** [review]
I don't really find `x == false` to be preferred over `!x`, but there is precedent in Zephyr, and it might be some MISRA-C thing.  So OK.

**Comment by jhedberg** [comment]
> I don't really find x == false to be preferred over !x, but there is precedent in Zephyr, and it might be some MISRA-C thing. So OK.

@pabigot actually bool types are (AFAIK) the only exception to this. For me it's not just allowed but recommended since it improves readability (IMO). @vikrant8051 your changes look good to me now except for this thing, could you please fix it and then we can get this finally merged? Thanks!

**Comment by vikrant8051** [comment]
@jhedberg @pabigot done !!

**Review by jhedberg — APPROVED** [review]

**Comment by vikrant8051** [comment]
@jhedberg could you please ask someone to execute PTS against entire APP after merging this PR ?

**Comment by jhedberg** [comment]
> @jhedberg could you please ask someone to execute PTS against entire APP after merging this PR?

@vikrant8051 we'll soon (next week) be entering the stabilization phase for Zephyr 1.14, during which this kind of stuff should be done. I can't guarantee we'll find someone to run these manual tests for your app - we have automated periodic (weekly, IIRC) runs for Mesh with tests/bluetooth/tester, but any other app requires manual running.

The best would be if you could just figure out how to get hold of a PTS USB dongle yourself.

**Comment by vikrant8051** [comment]
Please merge. And thanks for review.

**Pull request merged**
